Makefile

```diff
@@ -1,92 +1,141 @@
-	-e COVERALLS_REPO_TOKEN=$(COVERALLS_REPO_TOKEN) \
-	calico/test sh -c \
-	'pip install pykube && \
-	nosetests tests -c nose.cfg \
-	--with-xunit --xunit-file=/circle_output/output.xml; RC=$$?;\
-	[[ ! -z "$$COVERALLS_REPO_TOKEN" ]] && coveralls || true; exit $$RC'
+# Build the calico ipam plugin
+dist/calico-ipam: $(SRCFILES) vendor
+	CGO_ENABLED=0 go build -o dist/calico-ipam  \
+	-ldflags "-X main.VERSION=$(shell git describe --tags --dirty)" ipam/calico-ipam.go;
 
-clean:
-	-rm -f *.created
-	find . -name '*.pyc' -exec rm -f {} +
-	-sudo rm -rf dist
-	-docker run -v /var/run/docker.sock:/var/run/docker.sock -v /var/lib/docker:/var/lib/docker --rm martin/docker-cleanup-volumes
-	rm -f calico_cni/version.py
+.PHONY: test
+# Run the unit tests.
+test: dist/calico dist/calico-ipam run-etcd
+	# The tests need to run as root
+	sudo ETCD_IP=127.0.0.1 PLUGIN=calico GOPATH=$(GOPATH) $(shell which ginkgo)
+
+# Run the unit tests, watching for changes.
+test-watch: dist/calico dist/calico-ipam
+	# The tests need to run as root
+	sudo ETCD_IP=127.0.0.1 PLUGIN=calico GOPATH=$(GOPATH) $(shell which ginkgo) watch
+
+
+$(BUILD_CONTAINER_MARKER): Dockerfile
+	docker build -t $(BUILD_CONTAINER_NAME) .
+	touch $(BUILD_CONTAINER_MARKER)
+
+# Run the tests in a container. Useful for CI
+.PHONY: test-containerized
+test-containerized: dist/host-local run-etcd $(BUILD_CONTAINER_MARKER) build-containerized
+	docker run -ti --rm --privileged --net=host \
+	-e ETCD_IP=$(LOCAL_IP_ENV) \
+	-e PLUGIN=calico \
+	-v ${PWD}:/go/src/github.com/projectcalico/calico-cni:ro \
+	$(BUILD_CONTAINER_NAME) ginkgo
+
+# Run the build in a container. Useful for CI
+.PHONY: build-containerized
+build-containerized: $(BUILD_CONTAINER_MARKER) vendor
+	mkdir -p dist
+	docker run --rm \
+	-v ${PWD}:/go/src/github.com/projectcalico/calico-cni:ro \
+	-v ${PWD}/dist:/go/src/github.com/projectcalico/calico-cni/dist \
+	$(BUILD_CONTAINER_NAME) bash -c '\
+		make binary; \
+		chown -R $(shell id -u):$(shell id -u) dist'
 
-## Run etcd in a container. Generally useful.
+# Etcd is used by the tests
 run-etcd:
 	@-docker rm -f calico-etcd
 	docker run --detach \
 	--net=host \
-	--name calico-etcd quay.io/coreos/etcd:v2.2.2 \
+	--name calico-etcd quay.io/coreos/etcd:v2.3.6 \
 	--advertise-client-urls "http://$(LOCAL_IP_ENV):2379,http://127.0.0.1:2379,http://$(LOCAL_IP_ENV):4001,http://127.0.0.1:4001" \
 	--listen-client-urls "http://0.0.0.0:2379,http://0.0.0.0:4001"
 
-run-kubernetes-master: stop-kubernetes-master run-etcd  kubectl # binary dist/calicoctl
+# Install or update the tools needed for the static checks.
+.PHONY: update-tools
+update-tools:
+	go get -u github.com/Masterminds/glide
+	go get -u github.com/kisielk/errcheck
+	go get -u golang.org/x/tools/cmd/goimports
+	go get -u github.com/golang/lint/golint
+
+# Perform static checks on the code. The golint checks are allowed to fail, the others must pass.
+.PHONY: static-checks
+static-checks: vendor
```

Comment:
Seems like most repos have a `make release` type option to create a new release that builds the binaries and runs any required tests / checks.  I think we should have that to make releases easier and more repeatable.